l10n_it_account/models/account_group.py

```diff
@@ -17,29 +17,97 @@ class AccountGroup(models.Model):
         string="Balance sign",
     )
 
-    @api.constrains('account_ids')
-    def check_constrain_account_types(self):
-        error_msg = '\n'.join([
-            _("Cannot link accounts of different types to group '{}'."
-              .format(group.name))
-            for group in self
-            if len(group.account_ids.mapped('user_type_id').ids) > 1
-        ])
-        if error_msg:
-            raise ValidationError(error_msg)
+    @api.constrains('parent_id')
+    def check_parent_recursion(self):
+        for group in self:
+            try:
+                group.get_group_parents()
+            except ValidationError as err:
+                raise ValidationError(
+                    _("Can't set '{}' as parent for group '{}'."
+                      "\n{}")
+                    .format(group.parent_id.name_get()[0][-1],
+                            group.name_get()[0][-1],
+                            err.name)
+                )
+
+    @api.constrains('account_ids', 'parent_id')
+    def check_balance_sign_coherence(self):
+        """
+        Checks whether every group (plus parents and subgroups) have the same
+        balance sign. This is done by first retrieving every group's progenitor
+        and then checking, for each of them, the account types' for accounts
+        linked to such progenitor group and its subgroups.
+        """
+        # Force recursion check
+        self.check_parent_recursion()
+        if self.env.context.get('skip_check_balance_sign_coherence'):
+            return
+        done_group_ids = []
+        progenitor_ids = []
+        for group in self:
+            if group.id in done_group_ids:
```

Comment:
Credi che possiamo far rientrare questo if in un filtered sul self della riga precedente? Che ne pensi?

Reply:
@OpenCode non credo in realtà, perché `done_group_ids` viene inizializzato come una lista vuota che viene poi aggiornata ad ogni ciclo, quindi il controllo risulterebbe vano filtrando a monte del ciclo stesso